### DISCUSSION TITLE
Unable to find every sub definition in my perl module

### DISCUSSION THREAD

**Discussion by SteveWaring** [Camelcade/Perl5-IDEA#2871] [Q&A]
I've just installed Intellij Idea Community with the perl plugin. I've added my modules directory as a perl5 library. However, every subroutine call to my module is flagged as Unable to find. The code runs fine both externally and using the "play" button. How can I get rid of all these false errors? I have attached a screen shot. 
![Intellij](https://github.com/user-attachments/assets/1ae346e7-8e1a-402a-a204-abd43196a423)

**Comment by hurricup** [comment]
have you checked https://github.com/Camelcade/Perl5-IDEA/wiki/Subs-resolution-status ?
Because of nature of perl, subs may be defined/used in many ways and not all of them handled properly. So this question may be answered on case-by-case basis. In general - plugin does its best, but it won't ever be able to resolve everything correctly, because to do that it needs to run and simulate execution of all files in the project and plugin does not do that.

**Reply by SteveWaring** [reply]
Thanks. My main problem was that I had not refreshed the subs parsing. However, I still have problems with optionally imported subs. Eg:
use POSIX qw(strftime);
my $date = strftime($format, @when);

**Reply by hurricup** [reply]
Two things are important in such case: 
1. How sub is defined originally. If this is normal `sub strftime` or xs defined sub detected by deparser, it should be fine. If it is some tricky stuff defined using heavy perl magick - plugin may miss it.
2. How sub is exported. If this is simple using of `Exporter` and plain exporter array filling - should work fine. If there is more trickery (custom export, some fancy exporter array filling), plugin may miss it again. To solve this there are two ways for us to solve this:
- teach plugin new stuff (viable if approach used widely, not only in this particular package)
- hardcode exports in xml file (mostly used for core and popular modules)

User may write a pseudo-file assisting the plugin to understand what is going on (like auto-generated deparsed one).
But there is no point in the time when every perl user going to be 100% happy with how the plugin understands arbitrary perl code, because of the nature of perl and approach plugin uses to do its work.
